src/main/java/frc/robot/subsystems/shooter/ShooterDynamics.java

```diff
@@ -0,0 +1,115 @@
+
+  private static final double ROBOT_SIZE_RADIUS = Units.inchesToMeters(25.0);
+
+  public static Pair<Rotation2d, ShooterState> getRobotState(
+      Pose2d robotPose, Vector<N2> robotVel) {
+    return null;
+  }
+
+  // // TODO
+  // public static ShooterState calculateSpeaker(Vector<N2> robotVel, Pose2d robotPose) {
+  //   return null;
+  // }
+
+  // /**
+  //  * Calculates a scalar value of how much of a bias should be placed on a given pose to make a
+  // speaker shot
+  //  *
+  //  * @param robotPose the position to calculate for.
+  //  * @return bias coefficient.
+  //  */
+  // private double calculateSpeakerPoseBias(Pose2d robotPose, Vector<N2> robotVel) {
+  //
+  //
+  //
+  //
+  //
+  // }
+
+  public boolean isValidRobotState(Pose2d robotPose, Vector<N2> robotVel) {
+    // Ensure robot pose can legally shoot from there
+    if (robotPose.getX() + ROBOT_SIZE_RADIUS >= OTHER_ALLIANCE_WING_LINE) return false;
+
+    // Calculate the position of the pivot of the robot on the field
+    var pivotPose = getPivotPoseField(robotPose);
+
+    // Calculate the distance between the pivot of the shooter and the center mark of the speaker in
+    // two-dimensional space
+    var distToTarget =
+        Math.hypot(pivotPose.getX() - pivotPose.getX(), pivotPose.getY() - robotPose.getY());
+
+    // Ensure that the shooter could possibly make the shot with a reasonable trajectory
+    if (distToTarget > MAX_SHOOTER_DISTANCE) return false;
+
+    // Calculate if a note with lateral drift would make it into the speaker
+    double noteDriftLeft = 0.0; // TODO constant + function of velocity
+    double noteDriftRight = 0.0; // TODO constant + function of velocity
+
+    // TODO calculate note intersection points. Should allow for bank-shots on the inside sides of
+    // the speaker as well.
+    //  If the note cant make it in at all angles within the angle bound, return 0.
+
+    return true;
+  }
+
+  private static Pose3d getPivotPoseField(Pose2d robotPose) {
+    return new Pose3d(robotPose)
+        .transformBy(
+            new Transform3d(
+                Constants.Shooter.PIVOT_POSE.getTranslation(),
+                Constants.Shooter.PIVOT_POSE.getRotation()));
+  }
+
+  /**
+   * Represent the state of the shooter at a given time
+   *
+   * @param angle the angle of the shooter relative to the plane intersecting the erector parallel
+   *     to the ground.
+   * @param shooterVelocityMetersPerSecond the velocity of the shooter's flywheels in meters per
+   *     second. Positive values represent shooting the note out while negative values draw notes
+   *     in.
+   * @param kickupPercent scalar value representing the state of the kickup roller of the shooter.
+   *     Positive values represent pushing notes into the flywheels while negative values retract
+   *     the note into the shooter.
+   */
+  public record ShooterState(
+      Rotation2d angle, double shooterVelocityMetersPerSecond, double kickupPercent) {
```

Comment:
The kickup percentage shouldn't be controlled within the ShooterState class. This class should be reserved for setpoint style control and the kickup state should instead be moved directly into the shooter subsystem with methods to control it.

Reply:
Right away sir